### PR TITLE
prune slasher attestations

### LINKED ISSUE #7334: Activate attestation pruning in slasher db
# 🐞 Bug Report

### Description

Slasher disk space footprint is quite big (~300MB per day ~90gb for the whole weak subjectivity period) pruning old attestations (older then weak subjectivity should help confine the resources needed to run a slasher

### Has this worked before in a previous version?

was removed at some stage from the codebase

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

 Bug fix

**What does this PR do? Why is it needed?**
use prune attestation in slasher attestation store

old spanner implementation has been removed as well

**Which issues(s) does this PR fix?**

Fixes #7334

**Other notes for review**

### CODE REVIEW

File: slasher/db/kv/spanner.go [hunk 1]
```diff
@@ -260,8 +262,11 @@ func (db *Store) SaveValidatorEpochSpan(
 func (db *Store) SaveEpochSpansMap(ctx context.Context, epoch uint64, spanMap map[uint64]types.Span) error {
 	ctx, span := trace.StartSpan(ctx, "slasherDB.SaveEpochSpansMap")
 	defer span.End()
+	err := db.setObservedEpochs(ctx, epoch)
+	if err != nil {
```
Comment (0xKiwi) [inline slasher/db/kv/spanner.go:266]:
```suggestion
	if err := db.setObservedEpochs(ctx, epoch); err != nil {
```

File: slasher/db/kv/spanner.go [hunk 2]
```diff
@@ -232,8 +231,11 @@ func (db *Store) SaveValidatorEpochSpan(
 ) error {
 	ctx, traceSpan := trace.StartSpan(ctx, "slasherDB.SaveValidatorEpochSpan")
 	defer traceSpan.End()
+	err := db.setObservedEpochs(ctx, epoch)
+	if err != nil {
```
Comment (0xKiwi) [inline slasher/db/kv/spanner.go:235]:
```suggestion
	if err := db.setObservedEpochs(ctx, epoch); err != nil {
```

File: slasher/db/kv/spanner.go [hunk 3]
```diff
@@ -370,13 +375,19 @@ func (db *Store) findOrLoadEpochInCache(ctx context.Context, epoch uint64) (map[
...
 
-func setObservedEpochs(epoch uint64) {
+func (db *Store) setObservedEpochs(ctx context.Context, epoch uint64) error {
+	var err error
 	if epoch > highestObservedEpoch {
 		slasherHighestObservedEpoch.Set(float64(epoch))
 		highestObservedEpoch = epoch
+		// Prune block header history every PruneSlasherStoragePeriod epoch.
+		if highestObservedEpoch%params.BeaconConfig().PruneSlasherStoragePeriod == 0 {
+			err = db.PruneAttHistory(ctx, epoch, params.BeaconConfig().WeakSubjectivityPeriod)
```
Comment (0xKiwi) [inline slasher/db/kv/spanner.go:385]: Any reason not to handle the error immediately here?

File: slasher/db/kv/spanner.go (hunk 1 above)
Comment (shayzluf) [reply slasher/db/kv/spanner.go:266]: 10x

File: slasher/db/kv/spanner.go (hunk 2 above)
Comment (shayzluf) [reply slasher/db/kv/spanner.go:235]: 10x

File: slasher/db/kv/spanner.go (hunk 3 above)
Comment (shayzluf) [reply slasher/db/kv/spanner.go]: changed

File: slasher/db/kv/spanner_new.go [hunk 4]
```diff
@@ -103,3 +130,27 @@ func (db *Store) CacheLength(ctx context.Context) int {
 	log.Debugf("Span cache length %d", length)
 	return length
 }
+
+// EnableSpanCache used to enable or disable span map cache in tests.
+func (db *Store) EnableSpanCache(enable bool) {
+	db.spanCacheEnabled = enable
+}
+
+func (db *Store) setObservedEpochs(ctx context.Context, epoch uint64) error {
```
Comment (prestonvanloon) [inline slasher/db/kv/spanner_new.go:139]: Needs tests

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: slasher/db/kv/spanner_new.go (hunk 4 above)
Comment (shayzluf) [reply slasher/db/kv/spanner_new.go:139]: added tests

**Review by nisdas — COMMENTED** [review]
Why is the spanner being removed ?

**Comment by shayzluf** [comment]
@nisdas its the old implementation . its has not been used for vary long time

**Comment by shayzluf** [comment]
added to pr description

**Review by nisdas — APPROVED** [review]

**Pull request merged**